firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +187,960 @@ def interp_nullspace(I, nsp):
-            else:
-                self.pc.apply(x, y)
+            self.fdm_interp.mult(self.work_vec_y, y)
+            y.array_w[self.bc_nodes] = x.array_r[self.bc_nodes]
+        else:
+            self.pc.apply(x, y)
 
     def applyTranspose(self, pc, x, y):
-        dm = self._dm
-        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref):
-            if hasattr(self, "fdm_interp"):
-                self.fdm_interp.multTranspose(x, self.work_vec_y)
+        if hasattr(self, "fdm_interp"):
+            self.fdm_interp.multTranspose(x, self.work_vec_y)
+            with dmhooks.add_hooks(self._dm, self, appctx=self._ctx_ref):
                 self.pc.applyTranspose(self.work_vec_y, self.work_vec_x)
-                self.fdm_interp.mult(self.work_vec_x, y)
-                y.array_w[self.bc_nodes] = x.array_r[self.bc_nodes]
-            else:
-                self.pc.applyTranspose(x, y)
+            self.fdm_interp.mult(self.work_vec_x, y)
+            y.array_w[self.bc_nodes] = x.array_r[self.bc_nodes]
+        else:
+            self.pc.applyTranspose(x, y)
 
     def view(self, pc, viewer=None):
         super(FDMPC, self).view(pc, viewer)
         if hasattr(self, "pc"):
             viewer.printfASCII("PC to apply inverse\n")
             self.pc.view(viewer)
 
-    def assemble_fdm_op(self, V, J, bcs, appctx):
+    def destroy(self, pc):
+        objs = []
+        if hasattr(self, "A"):
+            objs.append(self.A)
+        if hasattr(self, "pc"):
+            objs.append(self.pc.getOperators()[-1])
+            objs.append(self.pc)
+        if hasattr(self, "submats"):
+            objs.extend(self.submats)
+        if hasattr(self, "work_mats"):
+            objs.extend(list(self.work_mats.values()))
+        if hasattr(self, "ises"):
+            objs.extend(self.ises)
+        for obj in objs:
+            if hasattr(obj, "destroy"):
+                obj.destroy()
+
+    @PETSc.Log.EventDecorator("FDMSetValues")
+    def set_values(self, A, Vrow, Vcol, addv, triu=False):
         """
-        Assemble the sparse preconditioner with cell-wise constant coefficients.
+        Assemble the stiffness matrix in the FDM basis using sparse reference
+        tensors and diagonal mass matrices.
 
-        :arg V: the :class:`~.FunctionSpace` of the form arguments
-        :arg J: the Jacobian bilinear form
-        :arg bcs: an iterable of boundary conditions on V
-        :arg appctx: the application context
+        :arg A: the :class:`PETSc.Mat` to assemble
+        :arg Vrow: the :class:`.FunctionSpace` test space
+        :arg Vcol: the :class:`.FunctionSpace` trial space
+        :arg addv: a `PETSc.Mat.InsertMode`
+        :arg triu: are we assembling only the upper triangular part?
+        """
+        def RtAP(R, A, P, result=None):
+            RtAP.buff = A.matMult(P, result=RtAP.buff)
+            return R.transposeMatMult(RtAP.buff, result=result)
+        RtAP.buff = None
```

Comment:
Same comment about function pointer attributes as above.

Reply:
replaced this function for a triple matrix product